// gatsby-plugin-breadcrumb/components/options-context.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.OptionsConsumer = exports.OptionsProvider = exports.OptionsContext = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

/* eslint-disable import/no-extraneous-dependencies */
var OptionsContext = _react["default"].createContext('Options');

exports.OptionsContext = OptionsContext;

var OptionsProvider = function OptionsProvider(_ref) {
  var children = _ref.children,
      _ref$useClassNames = _ref.useClassNames,
      useClassNames = _ref$useClassNames === void 0 ? false : _ref$useClassNames,
      _ref$useAutoGen = _ref.useAutoGen,
      useAutoGen = _ref$useAutoGen === void 0 ? false : _ref$useAutoGen;
  var options = {
    useClassNames: useClassNames,
    useAutoGen: useAutoGen
  };
  return _react["default"].createElement(OptionsContext.Provider, {
    value: options
  }, children);
};

exports.OptionsProvider = OptionsProvider;
var OptionsConsumer = OptionsContext.Consumer;
exports.OptionsConsumer = OptionsConsumer;
OptionsProvider.defaultProps = {
  useClassNames: false,
  useAutoGen: false
};
OptionsProvider.propTypes = {
  children: _propTypes["default"].node.isRequired,
  useClassNames: _propTypes["default"].bool,
  useAutoGen: _propTypes["default"].bool
};
