// gatsby-plugin-breadcrumb/components/useBreadcrumb.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _BreadcrumbContext = require("./BreadcrumbContext");

/* eslint-disable import/no-extraneous-dependencies */
var useBreadcrumb = function useBreadcrumb(_ref) {
  var location = _ref.location,
      crumbLabel = _ref.crumbLabel,
      crumbSeparator = _ref.crumbSeparator,
      _ref$crumbStyle = _ref.crumbStyle,
      crumbStyle = _ref$crumbStyle === void 0 ? {} : _ref$crumbStyle,
      _ref$crumbActiveStyle = _ref.crumbActiveStyle,
      crumbActiveStyle = _ref$crumbActiveStyle === void 0 ? {} : _ref$crumbActiveStyle;

  var _React$useContext = _react["default"].useContext(_BreadcrumbContext.BreadcrumbContext),
      crumbs = _React$useContext.crumbs,
      updateCrumbs = _React$useContext.updateCrumbs;

  if (!location || !crumbLabel) {
    return {
      crumbs: crumbs
    };
  }

  _react["default"].useEffect(function () {
    updateCrumbs({
      location: location,
      crumbLabel: crumbLabel,
      crumbSeparator: crumbSeparator,
      crumbStyle: crumbStyle,
      crumbActiveStyle: crumbActiveStyle
    });
  }, [location]);

  return {
    crumbs: crumbs
  };
};

var _default = useBreadcrumb;
exports["default"] = _default;
