// tsdx/dist/createRollupConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@babel/core");
const utils_1 = require("./utils");
const constants_1 = require("./constants");
const rollup_plugin_terser_1 = require("rollup-plugin-terser");
const rollup_plugin_babel_1 = tslib_1.__importDefault(require("rollup-plugin-babel"));
const rollup_plugin_commonjs_1 = tslib_1.__importDefault(require("rollup-plugin-commonjs"));
const rollup_plugin_json_1 = tslib_1.__importDefault(require("rollup-plugin-json"));
const rollup_plugin_replace_1 = tslib_1.__importDefault(require("rollup-plugin-replace"));
const rollup_plugin_node_resolve_1 = tslib_1.__importDefault(require("rollup-plugin-node-resolve"));
const rollup_plugin_sourcemaps_1 = tslib_1.__importDefault(require("rollup-plugin-sourcemaps"));
const rollup_plugin_typescript2_1 = tslib_1.__importDefault(require("rollup-plugin-typescript2"));
const replacements = [{ original: 'lodash', replacement: 'lodash-es' }];
const babelOptions = (format, target) => ({
    exclude: 'node_modules/**',
    extensions: [...core_1.DEFAULT_EXTENSIONS, 'ts', 'tsx'],
    passPerPreset: true,
    presets: [
        [
            require.resolve('@babel/preset-env'),
            {
                loose: true,
                modules: false,
                targets: target === 'node' ? { node: '8' } : undefined,
                exclude: ['transform-async-to-generator'],
            },
        ],
    ],
    plugins: [
        require.resolve('babel-plugin-annotate-pure-calls'),
        require.resolve('babel-plugin-dev-expression'),
        format !== 'cjs' && [
            require.resolve('babel-plugin-transform-rename-import'),
            { replacements },
        ],
        [
            require.resolve('babel-plugin-transform-async-to-promises'),
            { inlineHelpers: true, externalHelpers: true },
        ],
        [
            require.resolve('@babel/plugin-proposal-class-properties'),
            { loose: true },
        ],
    ].filter(Boolean),
});
// shebang cache map thing because the transform only gets run once
let shebang = {};
function createRollupConfig(format, env, opts) {
    return {
        // Tell Rollup the entry point to the package
        input: opts.input,
        // Tell Rollup which packages to ignore
        external: (id) => {
            if (id === 'babel-plugin-transform-async-to-promises/helpers') {
                return false;
            }
            return utils_1.external(id);
        },
        // Establish Rollup output
        output: {
            // Set filenames of the consumer's package
            file: `${constants_1.paths.appDist}/${utils_1.safePackageName(opts.name)}.${format}.${env}.js`,
            // Pass through the file format
            format,
            // Do not let Rollup call Object.freeze() on namespace import objects
            // (i.e. import * as namespaceImportObject from...) that are accessed dynamically.
            freeze: false,
            // Do not let Rollup add a `__esModule: true` property when generating exports for non-ESM formats.
            esModule: false,
            // Rollup has treeshaking by default, but we can optimize it further...
            treeshake: {
                // We assume reading a property of an object never has side-effects.
                // This means tsdx WILL remove getters and setters on objects.
                //
                // @example
                //
                // const foo = {
                //  get bar() {
                //    console.log('effect');
                //    return 'bar';
                //  }
                // }
                //
                // const result = foo.bar;
                // const illegalAccess = foo.quux.tooDeep;
                //
                // Punchline....Don't use getters and setters
                propertyReadSideEffects: false,
            },
            name: opts.name || utils_1.safeVariableName(opts.name),
            sourcemap: true,
            globals: { react: 'React', 'react-native': 'ReactNative' },
            exports: 'named',
        },
        plugins: [
            rollup_plugin_node_resolve_1.default({
                mainFields: [
                    'module',
                    'main',
                    opts.target !== 'node' ? 'browser' : undefined,
                ].filter(Boolean),
            }),
            format === 'umd' &&
                rollup_plugin_commonjs_1.default({
                    // use a regex to make sure to include eventual hoisted packages
                    include: /\/node_modules\//,
                }),
            rollup_plugin_json_1.default(),
            {
                // Custom plugin that removes shebang from code because newer
                // versions of bublé bundle their own private version of `acorn`
                // and I don't know a way to patch in the option `allowHashBang`
                // to acorn. Taken from microbundle.
                // See: https://github.com/Rich-Harris/buble/pull/165
                transform(code) {
                    let reg = /^#!(.*)/;
                    let match = code.match(reg);
                    shebang[opts.name] = match ? '#!' + match[1] : '';
                    code = code.replace(reg, '');
                    return {
                        code,
                        map: null,
                    };
                },
            },
            rollup_plugin_typescript2_1.default({
                typescript: require('typescript'),
                cacheRoot: `./.rts2_cache_${format}`,
                tsconfig: opts.tsconfig,
                tsconfigDefaults: {
                    compilerOptions: {
                        sourceMap: true,
                        declaration: true,
                        jsx: 'react',
                    },
                },
                tsconfigOverride: {
                    compilerOptions: {
                        target: 'esnext',
                    },
                },
            }),
            rollup_plugin_babel_1.default(babelOptions(format, opts.target)),
            rollup_plugin_replace_1.default({
                'process.env.NODE_ENV': JSON.stringify(env),
            }),
            rollup_plugin_sourcemaps_1.default(),
            // sizeSnapshot({
            //   printInfo: false,
            // }),
            env === 'production' &&
                rollup_plugin_terser_1.terser({
                    sourcemap: true,
                    output: { comments: false },
                    compress: {
                        keep_infinity: true,
                        pure_getters: true,
                        passes: 10,
                    },
                    ecma: 5,
                    toplevel: format === 'es' || format === 'cjs',
                    warnings: true,
                }),
        ],
    };
}
exports.createRollupConfig = createRollupConfig;
